sdks/python/apache_beam/runners/worker/statesampler_interface.py

```diff
@@ -0,0 +1,24 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from abc import ABC, abstractmethod
+
+
+class StateSamplerInterface(ABC):
+  @abstractmethod
+  def update_metric(self, typed_metric_name, value):
```

Comment:
It looks strange to a reader unfamiliar with this change why we are abstracting only one particular method of  StateSampler. Furthermore, we are reasoning about the logic how to accumulate the metrics in the stub, which might go out-of sync.

I wonder whether we can replace the statesampler reference in the subprocess with a unittest.mock.MagicMock, and then replay the calls returned by the_mock.mock_calls() on the real reference without introducing another class. As long as the mock doesn't need to return any values, this should work.  @robertwb do you think this would be too  hacky? Are there some concerns if the replay happens only after the do_fn.process() call finishes?

Reply:
That's an interesting idea, however I'd rather avoid dependencies on test classes like MagicMock. (It may also have issues with the pickling required here). I do think if we're going to introduce an interface here, we should list all the relevant methods and have the stub explicitly raise an error for the unsupported ones to clarify that it's a partial implementation.

Reply:
@robertwb do we need to apply the interface onto the cython implementation as well in this case? Does cython support inheriting from a python ABC?

Reply:
Discussed with @robertwb offline - cdef Cython classes cannot inherit from python classes, there is a limited support for multiple inheritance but comes with a potential performance hit, and this is performance sensitive code; we can leave the cython codepath alone and add a comment about the existence of the StateSampler interfaces defined in python codebase but not use them in Cythonized codepath.